internal/jobcontainers/job_test.go

```diff
@@ -0,0 +1,269 @@
+			return err
+		}
+		go proc.waitBackground(context.Background())
+	}
+	return nil
+}
+
+func cleanup(job *jobObject, procs []*JobProcess) {
+	job.close()
+	for _, proc := range procs {
+		if code, _ := proc.ExitCode(); code == -1 {
+			proc.Kill(context.Background())
+		}
+	}
+}
+
+func TestCreateAndTerminateJob(t *testing.T) {
+	job, err := createJobObject("test")
+	if err != nil {
+		t.Fatalf("failed to create job object: %s", err)
+	}
+
+	procs := createProcesses(2)
+
+	defer func() {
+		cleanup(job, procs)
+	}()
+
+	if err := startAndWait(job, procs); err != nil {
+		t.Fatalf("failed to start and wait on processes")
+	}
+
+	if err := job.terminate(); err != nil {
+		t.Fatalf("failed to terminate job object: %s", err)
+	}
+}
+
+func TestCreateAndShutdownJob(t *testing.T) {
+	job, err := createJobObject("test")
+	if err != nil {
+		t.Fatalf("failed to create job object: %s", err)
+	}
+
+	procs := createProcesses(2)
+
+	defer func() {
+		cleanup(job, procs)
+	}()
+
+	if err := startAndWait(job, procs); err != nil {
+		t.Fatalf("failed to start and wait on processes")
+	}
+
+	if err := job.shutdown(context.Background()); err != nil {
+		t.Fatalf("failed to shutdown job object: %s", err)
+	}
+}
+
+func TestJobPIDs(t *testing.T) {
+	job, err := createJobObject("test")
+	if err != nil {
+		t.Fatalf("failed to create job object: %s", err)
+	}
+
+	procs := createProcesses(2)
+
+	defer func() {
+		cleanup(job, procs)
+	}()
+
+	if err := startAndWait(job, procs); err != nil {
+		t.Fatalf("failed to start and wait on processes")
+	}
+
+	pidsMap := make(map[int]struct{})
+	for _, proc := range procs {
+		pidsMap[proc.Pid()] = struct{}{}
+	}
+
+	pids, err := job.pids()
+	if err != nil {
+		t.Fatalf("failed to get PIDs in job: %s", err)
+	}
+
+	if len(pids) != len(procs) {
+		t.Fatalf("number of PIDs in job incorrect")
+	}
+
+	for i := 0; i < len(pids); i++ {
+		if _, ok := pidsMap[int(pids[i])]; !ok {
+			t.Fatalf("PID not present in job object")
+		}
+	}
+
+	if err := job.shutdown(context.Background()); err != nil {
+		t.Fatalf("failed to shutdown job: %s", err)
+	}
+}
+
+func TestJobNotificationShutdown(t *testing.T) {
+	job, err := createJobObject("test")
+	if err != nil {
+		t.Fatalf("failed to create job object: %s", err)
+	}
+
+	procs := createProcesses(2)
+
+	defer func() {
+		cleanup(job, procs)
+	}()
+
+	if err := startAndWait(job, procs); err != nil {
+		t.Fatalf("failed to start and wait on processes")
+	}
+
+	go func() {
+		// Wait two seconds and then issue a shutdown. Should trigger the IOCP
+		// message if all the processes exit successfully.
+		time.Sleep(time.Second * 2)
+		if err := job.shutdown(context.Background()); err != nil {
+			fmt.Println(err)
+		}
+	}()
+
+	for {
+		code, err := job.pollIOCP()
+		if err != nil {
+			t.Fatalf("failed to poll IOCP: %s", err)
+		}
+
+		// No more processes in the job. Success state
+		if code == winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO {
+			return
+		}
+	}
```

Comment:
Not super important, but maybe we should also include a context with a timeout and cancel the function if we haven't received the anticipated code in 30 seconds or something? I'm not sure what the other responses look like though

Reply:
We only check for all processes exiting so we have no way of knowing when this is going to happen to be able to set any timeout. Could be 1 second after the jobs created, could be next month :P

Reply:
Yeah but since this is test code, I assume if it doesn't exit in like 30 seconds we should just return an error saying so instead of waiting until next month lol

Reply:
Oh I'm an idiot, I didn't read what file this was from I thought it was the main IOCP loop for the actual implementation. Yes agreed on the timeout

Reply:
Done